### PR TITLE
extend petsc variants by mumps, complex and debug, fix headers in superlu_dist, add variant to hypre

### PR DESCRIPTION
**Pull request by davydden**

### CODE REVIEW

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 1]
```diff
@@ -14,6 +14,8 @@ class Hypre(Package):
 
     # hypre does not know how to build shared libraries on Darwin
     variant('shared', default=sys.platform!='darwin', description="Build shared library version (disables static library)")
+    # SuperluDist have conflicting headers with those in Hypre
+    variant('internal-superlu', default=True, description="Use internal Superlu routines")
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/hypre/package.py]: I think the sense of this should be inverted.  `+superlu` should build with an external `superlu`, to mirror what other packages do, and you can make it off by default.  If it's off, petsc should use the internal `superlu` routines.

Comment (davydden) [inline var/spack/repos/builtin/packages/hypre/package.py]: i think you mixed a bit packages. If I understand you correctly, here, in `hypre` we do `+superlu` (build itnernal superlu). It does **not** (AFAIK) build against external one (thus possible collision in headers). Now another package, `superlu_dist` (!= superlu), also has those headers from `superlu` (serial version). Finally, `petsc` can be build against both `hypre` (preconditioner) and `superlu_dist` (direct parallel solver).

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 2]
```diff
@@ -40,7 +44,13 @@ class Petsc(Package):
 
     depends_on('hdf5+mpi', when='+hdf5+mpi')
     depends_on('parmetis', when='+metis+mpi')
-    depends_on('hypre',    when='+hypre+mpi')
+    # Hypre does not support comple numbers.
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/petsc/package.py]: *complex

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 1 above)
Comment (tgamblin) [inline var/spack/repos/builtin/packages/hypre/package.py]: Sorry, I *did* mean `hypre` above.  This makes sense -- `internal-superlu` is probably fine as it is currently.  I should  ask Ulrike and Rob about this -- currently they're all trying to collaborate on a release of the [`xSDK`](https://ideas-productivity.org/focus-a…), which includes `hypre`, `petsc`, `superlu`, `superlu_dist`, etc.  I'm curious if this option is going to stay the same.

Comment (davydden) [inline var/spack/repos/builtin/packages/hypre/package.py]: good to know!

p.s. here is the description in config
```
--with-superlu          Use internal SuperLU routines.
```
Hypre also supports CMake installation and there don't even ported this option, i.e. CMakeList contains
```
#   --with-examples         Build example codes.
#   --with-fei              Use internal FEI routines.
#   --with-superlu          Use internal SuperLU routines.
#   --with-mli              Use MLI
```
and that's the only mentioning of SuperLU there.

**Pull request merged**